### PR TITLE
Implemented uv_run2 (adding flags to uv_run)

### PR DESCRIPTION
**Pull request by saghul**
Allows for running the event loop in 3 modes:
- default: loop runs until the refcount drops to zero
- once: poll for events only once and block until one is handled
- nowait: poll for events only once but don't block if there are
  no pending events

The patch contains the Unix implementation, I'll do the Windows one after some initial feedback.

I thought changing uv_run could be a bit too much, so the new functionality is implemented as uv_run2. I did remove uv_run_once, since it's very straightforward to mock with a define, for software which runs on 0.8 and 0.9.

Related issue: #339

/cc @bnoordhuis @piscisaureus

### CODE REVIEW

**Comment by satoruhiga** [comment]
This pull request is still alive?
I want to use libuv in a single thread run loop.. (e.g. OpenGL application)

**Comment by saghul** [comment]
This will be reworked soon, once libev is removed from libuv.

**Comment by satoruhiga** [comment]
OK, Thanks! I'm really looking forward!

**Comment by txdv** [comment]
libev is removed from libuv, go implement it!

**Comment by saghul** [comment]
Ok, I rebased the code to work with latest master. Mr Travis says it's all good :-) This is just the unix side if of things, care to review it @bnoordhuis? I'll go for the windows side afterwards.

Btw, should we keep uv_run_once around? Perhaps with that new UV_DEPRECATED?

File: include/uv.h [hunk 1]
```diff
@@ -244,12 +251,18 @@
...
+ *    zero. Always returns zero.
+ *  - UV_RUN_ONCE: Poll for new events once. Note that this function blocks if
+ *    there are no pending events. Returns zero when done (no active handles
+ *    or requests left), or non-zero if more events are expected (meaning you
+ *    should run the event loop again sometime in the future).
+ *  - UV_RUN_NOWAIT: Poll for new events once but don't block if there are no
+ *    pending events.
  */
-UV_EXTERN int uv_run_once(uv_loop_t*);
+UV_EXTERN int uv_run2(uv_loop_t*, uv_run_flags flags);
```
Comment (bnoordhuis) [inline include/uv.h]: I suggest renaming `uv_run_flags flags` to `uv_run_mode mode`. 'flags' suggests it's possible to OR them together.

File: src/unix/core.c [hunk 2]
```diff
@@ -267,27 +267,25 @@ int uv_backend_timeout(const uv_loop_t* loop) {
...
+    uv__run_timers(loop);
+    uv__run_idle(loop);
+    uv__run_prepare(loop);
+    uv__run_pending(loop);
+    uv__io_poll(loop, (flags & UV_RUN_NOWAIT ? 0 : uv_backend_timeout(loop)));
+    uv__run_check(loop);
+    uv__run_closing_handles(loop);
+    r = uv__has_active_handles(loop) || uv__has_active_reqs(loop);
+  } while (r && !(flags & (UV_RUN_ONCE | UV_RUN_NOWAIT)));
+  return (flags & UV_RUN_DEFAULT ? 0 : r);
```
Comment (bnoordhuis) [inline src/unix/core.c]: Superfluous check? r should always be zero with UV_RUN_DEFAULT, right?

Comment (saghul) [inline src/unix/core.c]: Actually, yes. All tests assert after uv_run returns, so we'll catch problems if something goes wrong anyway.

**Comment by bnoordhuis** [comment]
Patch for uv.gyp:

``` diff
diff --git a/uv.gyp b/uv.gyp
index 9368329..2eda03f 100644
--- a/uv.gyp
+++ b/uv.gyp
@@ -275,6 +275,7 @@
         'test/test-poll-close.c',
         'test/test-process-title.c',
         'test/test-ref.c',
+        'test/test-run-nowait.c',
         'test/test-run-once.c',
         'test/test-semaphore.c',
         'test/test-shutdown-close.c',
```

**Comment by saghul** [comment]
Updated addressing comments by @bnoordhuis, thanks!

**Comment by bnoordhuis** [comment]
LGTM. Will you be adding the Windows side?

**Comment by saghul** [comment]
Yes, I'm on it, though I'm a bit out of my confort zone ;-)

File: src/win/core.c [hunk 3]
```diff
@@ -252,57 +252,47 @@ static void uv_poll_ex(uv_loop_t* loop, int block) {
...
+
+  if (pGetQueuedCompletionStatusEx)
+    poll = &uv_poll_ex;
+  else:
+    poll = &uv_poll;
+
+  r = UV_LOOP_ALIVE(loop);
+  while (r) {
+    uv_update_time((loop));
+    uv_process_timers((loop));
```
Comment (bnoordhuis) [inline src/win/core.c]: Extraneous parentheses.

**Comment by saghul** [comment]
Added Windows counterpart now. @piscisaureus can you please review?

PD: @bnoordhuis I removed the extra parens.

**Comment by piscisaureus** [comment]
Thanks @saghul, fixed and landed in 0820be7.

**Comment by saghul** [comment]
Awesome, thanks!

On Tuesday, December 18, 2012, Bert Belder wrote:

> Thanks @saghul https://github.com/saghul, fixed and landed in 0820be7https://github.com/joyent/libuv/commit/0820be7
> .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/joyent/libuv/pull/535#issuecomment-11490802.

## 

/Saúl
http://saghul.net | http://sipdoc.net

**Pull request closed without merging**